pkg/pipeline/pipeline_builder.go

```diff
@@ -0,0 +1,335 @@
+		case StageExtract:
+			pEntry.Extractor, err = getExtractor(*params)
+		case StageEncode:
+			pEntry.Encoder, err = getEncoder(*params)
+		case StageWrite:
+			pEntry.Writer, err = getWriter(*params)
+		default:
+			err = fmt.Errorf("invalid stage type: %v", pEntry.stageType)
+		}
+		if err != nil {
+			return err
+		}
+		b.pipelineEntryMap[stage.Name] = &pEntry
+		b.pipelineStages = append(b.pipelineStages, &pEntry)
+		log.Debugf("pipeline = %v", b.pipelineStages)
+	}
+	log.Debugf("pipeline = %v", b.pipelineStages)
+	return nil
+}
+
+// reads the configured Go stages and connects between them
+func (b *builder) build() (*Pipeline, error) {
+	if err := b.readStages(); err != nil {
+		return nil, err
+	}
+
+	for _, connection := range b.configStages {
+		if connection.Name == "" || connection.Follows == "" {
+			// ignore entries that do not represent a connection
+			continue
+		}
+		// instantiates (or loads from cache) the destination node of a connection
+		dstEntry, ok := b.pipelineEntryMap[connection.Name]
+		if !ok {
+			return nil, fmt.Errorf("unknown pipeline stage: %s", connection.Name)
+		}
+		dstNode, err := b.getStageNode(dstEntry)
+		if err != nil {
+			return nil, err
+		}
+		dst, ok := dstNode.(node.Receiver)
+		if !ok {
+			return nil, fmt.Errorf("stage %q of type %q can't receive data",
+				connection.Name, dstEntry.stageType)
+		}
+		// instantiates (or loads from cache) the source node of a connection
+		srcEntry, ok := b.pipelineEntryMap[connection.Follows]
+		if !ok {
+			return nil, fmt.Errorf("unknown pipeline stage: %s", connection.Follows)
+		}
+		srcNode, err := b.getStageNode(srcEntry)
+		if err != nil {
+			return nil, err
+		}
+		src, ok := srcNode.(node.Sender)
+		if !ok {
+			return nil, fmt.Errorf("stage %q of type %q can't send data",
+				connection.Name, dstEntry.stageType)
+		}
+		log.Debugf("connecting stages: %s --> %s", connection.Follows, connection.Name)
+
+		// connects source and destination node, and catches any panic from the Go-Pipes library.
+		var catchErr error
+		func() {
+			defer func() {
+				if msg := recover(); msg != nil {
+					catchErr = fmt.Errorf("%q and %q stages haven't compatible input/outputs: %v",
+						connection.Follows, connection.Name, msg)
+				}
+			}()
+			src.SendsTo(dst)
+		}()
+		if catchErr != nil {
+			return nil, catchErr
+		}
+	}
+
+	if len(b.startNodes) == 0 {
+		return nil, errors.New("no ingesters have been defined")
+	}
+	if len(b.terminalNodes) == 0 {
+		return nil, errors.New("no writers have been defined")
+	}
+	return &Pipeline{
+		startNodes:     b.startNodes,
+		terminalNodes:  b.terminalNodes,
+		pipelineStages: b.pipelineStages,
+	}, nil
+}
+
+func (b *builder) getStageNode(pe *pipelineEntry) (interface{}, error) {
+	if stg, ok := b.createdStages[pe.configStage.Name]; ok {
+		return stg, nil
+	}
+	var stage interface{}
+	// TODO: modify all the types' interfaces to not need to write loops here, the same
+	// as we do with Ingest
```

Comment:
I assume this is to be done in the next PR?

Reply:
I was considering to do it in yet another PR, after Goflow Kube is integrated with Flowlows Pipeline, we verify that works and then can cleanup some TODOs.